src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1407,6 +1407,34 @@ void Session::cancel_resumption_delay()
 }
 
 
+void Session::gather_pending_compensating_writes(util::Span<Changeset> changesets,
+                                                 std::vector<ProtocolErrorInfo>* out)
+{
+    if (m_pending_compensating_write_errors.empty() || changesets.empty()) {
+        return;
+    }
+
+#ifdef REALM_DEBUG
+    REALM_ASSERT_DEBUG(
+        std::is_sorted(m_pending_compensating_write_errors.begin(), m_pending_compensating_write_errors.end(),
+                       [](const ProtocolErrorInfo& lhs, const ProtocolErrorInfo& rhs) {
+                           return lhs.compensating_write_server_version < rhs.compensating_write_server_version;
+                       }));
+#endif
+
+    while (!m_pending_compensating_write_errors.empty() &&
+           m_pending_compensating_write_errors.front().compensating_write_server_version <=
+               changesets.back().version) {
+        auto& cur_error = m_pending_compensating_write_errors.front();
+        if (cur_error.compensating_write_server_version >= changesets.front().version) {
```

Comment:
Nice clean-up. Is this check needed? Kiro mentioned we are guaranteed to receive a changeset at compensating_write_server_version, so maybe the check should be an assertion instead